src/backend/distributed/transaction/transaction_management.c

```diff
@@ -545,3 +553,17 @@ SwallowErrors(void (*func)())
 	}
 	PG_END_TRY();
 }
+
+
+/*
+ * IsMultiStatementTransaction determines whether the current statement is
+ * part of a bigger multi-statement transaction. This is the case when the
+ * statement is wrapped in a transaction block (comes after BEGIN), or it
+ * is called from a stored procedure or function.
+ */
+bool
+IsMultiStatementTransaction(void)
+{
+	return IsTransactionBlock() || StoredProcedureLevel > 0 || (FunctionCallLevel > 0 &&
```

Comment:
I know what this disjunction means right now, but I'm not sure I'll know in a year or so. I'm struggling to think of a more comprehensible way to write it; the following is very verbose, but I feel it'll remain scrutable in the future?

```c
if (IsTransactionBlock())
{
    /* we're in a BEGIN ... END block, could have multi-stmt */
    return true;
}
else if (StoreProcedureLevel > 0)
{
    /* in a stored procedure */
    return true;
}
else if (FunctionCallLevel > 0)
{
    /* in a language-handler function call, open a transaction if configured to do so */
    return FunctionOpensTransactionBlock;
}
else
{
    /* apparently not in any sort of multi-stmt txn */
    return false;
}
```

The above seems a little crazy, but I do worry about that long boolean thing aging well.

Reply:
Makes sense. Will change it